src/index/index.c

```diff
@@ -11,80 +11,147 @@
 #include "../query_ctx.h"
 #include "../util/rmalloc.h"
 #include "../datatypes/point.h"
-#include "../graph/graphcontext.h"
-#include "../graph/entities/node.h"
-#include "../graph/rg_matrix/rg_matrix_iter.h"
+#include "../datatypes/vector.h"
 
 #include <stdatomic.h>
 
+// returns a type aware name for a given field
+// e.g.
+// "range:age" for a range field named "age"
+// "vector:location" for a vector field named "location"
+// "name" for a full text field named "name"
+void Index_TypedFieldName
+(
+	char *type_aware_name,  // [out] type aware name
+	IndexFieldType t,       // field type
+	const char *name        // field name
+) {
+	ASSERT(name != NULL);
+	ASSERT(type_aware_name != NULL);
+	ASSERT(t == INDEX_FLD_FULLTEXT ||
+		   t == INDEX_FLD_RANGE    ||
+		   t == INDEX_FLD_VECTOR);
+
+	if(t == INDEX_FLD_FULLTEXT) {
+		// maintain original name for full text fields
+		strcpy(type_aware_name, name);
+	} else if(t == INDEX_FLD_RANGE) {
+		// prefix range field name with "range:"
+		sprintf(type_aware_name, "range:%s", name);
+	} else if(t == INDEX_FLD_VECTOR) {
+		// prefix vector field name with "vector:"
+		sprintf(type_aware_name, "vector:%s", name);
+	} else {
+		assert(false && "unexpected field type");
+	}
+}
+
+// index structure
 struct _Index {
 	char *label;                   // indexed label
 	int label_id;                  // indexed label ID
 	IndexField *fields;            // indexed fields
 	char *language;                // language
 	char **stopwords;              // stopwords
 	GraphEntityType entity_type;   // entity type (node/edge) indexed
-	IndexType type;                // index type exact-match / fulltext
 	RSIndex *rsIdx;                // RediSearch index
 	uint _Atomic pending_changes;  // number of pending changes
 };
 
-static void _Index_ConstructFullTextStructure
+// merge field 'b' into 'a'
+static void _Index_MergeFields
+(
+	IndexField *a,
+	const IndexField *b
+) {
+	ASSERT(a != NULL);
+	ASSERT(b != NULL);
+	ASSERT((a->type & b->type) == 0);
+
+	// merge type
+	a->type |= b->type;
+
+	// merge options
+	if(b->type & INDEX_FLD_FULLTEXT) {
```

Comment:
move to switch case

Reply:
I don't believe I can as `INDEX_FLD_RANGE` isn't an enum value but a macro oring 3 different enum values.
`switch case` can't have `INDEX_FLD_RANGE` as one of its cases